docs/plugin-authoring-evaluator.md

```diff
@@ -0,0 +1,380 @@
+  typeof DeliciousnessDetectionResponseSchema
+>;
+
+/**
+ * Score an individual test case for delciousness.
+ */
+export async function deliciousnessScore<
+  CustomModelOptions extends z.ZodTypeAny,
+>(
+  judgeLlm: ModelArgument<CustomModelOptions>,
+  dataPoint: BaseDataPoint,
+  judgeConfig?: CustomModelOptions
+): Promise<Score> {
+  const d = dataPoint;
+  try {
+    // Validate the input has required fields
+    if (!d.output) {
+      throw new Error('Output is required for Funniness detection');
+    }
+
+    //Hydrate the prompt
+    const promptTemplate = Handlebars.compile(QUESTION_GEN_PROMPT);
+    const finalPrompt = promptTemplate({
+      output: d.output,
+    });
+
+    // Call the LLM to generate an evaluation result
+    const response = await generate({
+      model: judgeLlm,
+      prompt: finalPrompt,
+      config: judgeConfig,
+    });
+
+    // Parse the output
+    const parsedResponse = extractJson<DeliciousnessDetectionResponse>(
+      response.text()
+    );
+    if (!parsedResponse) {
+      throw new Error(`Unable to parse evaluator response: ${response.text()}`);
+    }
+
+    // Return a scored response
+    return {
+      score: parsedResponse.verdict,
+      details: { reasoning: parsedResponse.reason },
+    };
+  } catch (err) {
+    console.debug(
+      `Deliciousness evaluation failed with error ${err} for sample ${JSON.stringify(
+        d
+      )}`
+    );
+    throw err;
+  }
+}
+```
+
+#### Define the evaluator action
+
+The final step is to write a function that defines the evaluator action itself.
+
+```ts
+/**
+ * Create the Deliciousness evaluator action.
+ */
+export function createDeliciousnessEvaluator<
+  ModelCustomOptions extends z.ZodTypeAny,
+>(
+  judge: ModelReference<ModelCustomOptions>,
+  judgeConfig: z.infer<ModelCustomOptions>
+): EvaluatorAction {
+  return defineEvaluator(
+    {
+      name: `myAwesomeEval/deliciousness`,
+      displayName: 'Deliciousness',
+      definition: 'Determines if output is considered delicous.',
+    },
+    async (datapoint: BaseDataPoint) => {
+      const score = await deliciousnessScore(judge, datapoint, judgeConfig);
+      return {
+        testCaseId: datapoint.testCaseId,
+        evaluation: score,
+      };
+    }
+  );
+}
+```
+
+### Non-LLM-based Evaluators
+
+An non-LLM-based evaluator is any function used evaluate the input, context, or output of your genAI feature.
+
+Non-LLM-based evaluators in Genkit are made up of 2 components:
+
+- A scoring function
+- An evaluator action
+
+#### Define the scoring function
+
+Just like the LLM-based evaluator, define the scoring function. In this case, the scoring function does not need to know about the judge LLM or its config.
+
+```ts
+const US_PHONE_REGEX =
+  /^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4}$/i;
+
+/**
+ * Scores whether an individual datapoint matches a US Phone Regex.
+ */
+export async function usPhoneRegexScore(
+  dataPoint: BaseDataPoint
```

Comment:
It feels like it's better to provide an example where we use strongly typed datapoints, rather than the BaseDataPoint. WDYT?

Reply:
I'm going to do this in a followup if that's cool. I think it would be rad to show that, but it's also causing me a bit of trouble right now.